andela/haven-ah-backend
# #161342857 Create endpoint to report an article

#### What does this PR do?
Create endpoint to report an article. **POST** `/api/v1/articles/:slug/complaints`.

#### Description of Task to be completed
- add function to create a complaint.
- add the endpoint to post a complaint.
- abstract the function to check if an article exists.
- add a function to wrap controller functions in a try catch.

#### How should this be manually tested?
- on the terminal run git clone https://github.com/andela/haven-ah-backend.git.
- run git checkout `ft-post-complaints-161342857`.
- ensure that PostgreSQL is running on your local machine.
- run `npm install` to install the npm packages.
- run `npm test` to run the unit and integration tests.

To test on postman
- run npm run start-dev and open up postman.

To test the routes above, the user needs to be authenticated.
- signup (`localhost:5000/api/v1/users/signup`) and use email and password to log in (`localhost:5000/api/v1/users/signin`). Copy token and pass to the x-access-token header to test the endpoint.

Request Body should contain
```gherkin
{
  complaintType: One of ['Rules Violation', 'Abuse', 'Plagiarism', 'Others'],
  complaintBody:  "Report plagiarism",
}
```
Sample Responses
If user is the author of the article
```gherkin
{
  status: 400,
  message: "Sorry, you can not raise a complaint against your article."
}
```

if post is successful
```gherkin
{
  status: 201,
  message:  "You have logged a complaint on this article."
}
```

#### Screenshots if applicable
<img width="1098" alt="screen shot 2018-11-07 at 6 11 33 pm" src="https://user-images.githubusercontent.com/29674347/48147932-db8b4f00-e2b8-11e8-9e29-bf872491887f.png">

#### What are the relevant pivotal tracker stories?
#161342857
